File: system-test/typescript/src/testkit/dsl/core/usecase/UseCaseDsl.ts
import type { AppContext, ChannelMode } from '../scenario/app-context.js';
import { ShopDsl } from './shop/ShopDsl.js';
import { ClockDsl } from './external/clock/ClockDsl.js';
import { ErpDsl } from './external/erp/ErpDsl.js';
import { TaxDsl } from './external/tax/TaxDsl.js';

export class UseCaseDsl {
  private _shopDsl?: ShopDsl;
  private _clockDsl?: ClockDsl;
  private _erpDsl?: ErpDsl;
  private _taxDsl?: TaxDsl;

  constructor(private readonly app: AppContext) {}

  shop(mode?: ChannelMode): ShopDsl {
    if (!this._shopDsl || mode) {
      this._shopDsl = new ShopDsl(this.app.shop(mode));
    }
    return this._shopDsl;
  }

  clock(): ClockDsl {
    if (!this._clockDsl) {
      this._clockDsl = new ClockDsl(this.app.clockDriver);
    }
    return this._clockDsl;
  }

  erp(): ErpDsl {
    if (!this._erpDsl) {
      this._erpDsl = new ErpDsl(this.app.erpDriver);
    }
    return this._erpDsl;
  }

  tax(): TaxDsl {
    if (!this._taxDsl) {
      this._taxDsl = new TaxDsl(this.app.taxDriver);
    }
    return this._taxDsl;
  }

  async close(): Promise<void> {
    await this.app.closeAll();
  }
}

File: system/multitier/backend-clean-java/src/testSupport/java/com/mycompany/myshop/backend/testkit/dsl/core/usecase/UseCaseDsl.java
package com.mycompany.myshop.backend.testkit.dsl.core.usecase;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mycompany.myshop.backend.testkit.driver.port.MyShopDriver;
import com.mycompany.myshop.backend.testkit.driver.port.external.clock.ClockDriver;
import com.mycompany.myshop.backend.testkit.driver.port.external.erp.ErpDriver;
import com.mycompany.myshop.backend.testkit.driver.adapter.sut.SutClockReader;
import com.mycompany.myshop.backend.testkit.driver.adapter.sut.SutErpReader;
import com.mycompany.myshop.backend.testkit.driver.adapter.sut.SutTaxReader;
import com.mycompany.myshop.backend.testkit.driver.port.external.tax.TaxDriver;
import com.mycompany.myshop.backend.testkit.dsl.core.usecase.external.clock.ClockDsl;
import com.mycompany.myshop.backend.testkit.dsl.core.usecase.external.erp.ErpDsl;
import com.mycompany.myshop.backend.testkit.dsl.core.usecase.external.tax.TaxDsl;

public class UseCaseDsl {

    private final MyShopDsl myShop;
    private final ErpDsl erp;
    private final TaxDsl tax;
    private final ClockDsl clock;
    private final SutErpReader sutErp;
    private final SutTaxReader sutTax;
    private final SutClockReader sutClock;

    public UseCaseDsl(
            MyShopDriver backendDriver,
            ObjectMapper objectMapper,
            ErpDriver erpStubDriver,
            TaxDriver taxStubDriver,
            ClockDriver clockStubDriver,
            SutErpReader sutErp,
            SutTaxReader sutTax,
            SutClockReader sutClock) {
        this.myShop = new MyShopDsl(backendDriver, objectMapper);
        this.erp = new ErpDsl(erpStubDriver);
        this.tax = new TaxDsl(taxStubDriver);
        this.clock = new ClockDsl(clockStubDriver);
        this.sutErp = sutErp;
        this.sutTax = sutTax;
        this.sutClock = sutClock;
    }

    public MyShopDsl myShop() {
        return myShop;
    }

    public ErpDsl erp() {
        return erp;
    }

    public TaxDsl tax() {
        return tax;
    }

    public ClockDsl clock() {
        return clock;
    }

    public SutErpReader sutErp() {
        return sutErp;
    }

    public SutTaxReader sutTax() {
        return sutTax;
    }

    public SutClockReader sutClock() {
        return sutClock;
    }
}
